src/passes/MemoryCopyFillLowering.cpp

```diff
@@ -0,0 +1,256 @@
+/*
+ * Copyright 2024 WebAssembly Community Group participants
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "ir/names.h"
+#include "pass.h"
+#include "wasm-builder.h"
+#include "wasm.h"
+
+// Replace memory.copy and memory.fill with a call to a function that
+// implements the same semantics.
+
+namespace wasm {
+struct MemoryCopyFillLowering
+  : public WalkerPass<PostWalker<MemoryCopyFillLowering>> {
+  bool needsMemoryCopy = false;
+  bool needsMemoryFill = false;
+  Name memCopyFuncName;
+  Name memFillFuncName;
+
+  void visitMemoryCopy(MemoryCopy* curr) {
+    assert(curr->destMemory ==
+           curr->sourceMemory); // multi-memory not supported.
+    Builder builder(*getModule());
+    replaceCurrent(builder.makeCall(
+      "__memory_copy", {curr->dest, curr->source, curr->size}, Type::none));
+    needsMemoryCopy = true;
+  }
+
+  void visitMemoryFill(MemoryFill* curr) {
+    Builder builder(*getModule());
+    replaceCurrent(builder.makeCall(
+      "__memory_fill", {curr->dest, curr->value, curr->size}, Type::none));
+    needsMemoryFill = true;
+  }
+
+  void run(Module* module) override {
+    if (!module->features.hasBulkMemory()) {
+      return;
+    }
+    if (module->features.hasMemory64() || module->features.hasMultiMemory()) {
+      Fatal()
+        << "Memory64 and multi-memory not supported by memory.copy lowering";
+    }
+
+    // Check for the presence of any passive data or table segments.
+    for (auto& segment : module->dataSegments) {
+      if (segment->isPassive) {
+        Fatal() << "memory.copy lowering should only be run on modules with "
+                   "no passive segments";
+      }
+    }
+    for (auto& segment : module->elementSegments) {
+      if (!segment->table.is()) {
+        Fatal() << "memory.copy lowering should only be run on modules with"
+                   " no passive segments";
+      }
+    }
+
+    // In order to introduce a call to a function, it must first exist, so
+    // create an empty stub.
+    Builder b(*module);
```

Comment:
```suggestion
    Builder builder(*module);
```
Though maybe it's ok to differ from our usual convention if this is mentioned a huge number of times...

Reply:
Yeah, it would push out the line lengths quite a bit, and probably result in even more lines.